semantics/executable-spec/src/Data/Coders.hs

```diff
@@ -320,56 +421,160 @@ decodE x = decodeCount x 0
+applyField f seen name = do
+  tag <- decodeWord
+  if Set.member tag seen
+     then duplicateKey name tag
+     else case f tag of
+            Field update d -> do v <- decode d; pure (update v,insert tag seen)
+
+-- | Decode a Map Block of key encoded data for type t
+--   given a function that picks the right box for a given key, and an
+--   initial value for the record (usually starts filled with default values).
+--   The Block can be either len-encoded or block-encoded.
+
+getSparseBlock :: Maybe Int -> t -> (Word -> Field t) -> Set Word -> String -> Decoder s (t,Set Word)
+getSparseBlock (Just 0) initial _pick seen _name = pure(initial,seen)
+getSparseBlock (Just n) initial pick seen name = do
+   (transform,seen2) <- applyField pick seen name
+   getSparseBlock (Just(n-1)) (transform initial) pick seen2 name
+getSparseBlock Nothing initial pick seen name =
+   decodeBreakOr >>= \case
+             True -> pure(initial,seen)
+             False -> do (transform,seen2) <- applyField pick seen name
+                         getSparseBlock Nothing (transform initial) pick seen2 name
+
+-- ======================================================
+-- (Decode 'Closed) and (Decode 'Thin) are applicative
+-- (Decode 'Nary) is not applicative since there is no
+-- (Applys 'Nary 'Nary) instance. And there should never be one.
 
 instance Functor (Decode w) where
   fmap f (Map g x) = Map (f . g) x
   fmap f x = Map f x
 
+instance Applicative (Decode 'Closed) where
+  pure x = Emit x
+  f <*> x = ApplyD f x
+
+instance Applicative (Decode 'Thin) where
+  pure x = Emit x
+  f <*> x = ApplyD f x
+
 -- ===========================================================================================
--- These functions are the dual analogs to
--- Shelley.Spec.Ledger.Serialization(decodeList, decodeSeq, decodeStrictSeq, decodeSet)
--- It is not well documented how to use encodeFoldable.
--- They are provided here as compatible pairs for use with the (E x) and (D x) constructors
--- of the Encode and Decode types. (E encodeList xs) and (D (decodeList fromCBOR)) should be duals.
+-- A Dual pairs an Encoding and a Decoder with a roundtrip property.
+-- They are used with the (E and D) constructors of Encode and Decode
+-- If you are trying to code something not in the CBOR classes
+-- or you want something not traditional, make you own Dual and use E or D
+
+-- data Dual t = Dual (t -> Encoding) (forall s . Decoder s t)
+
+dualList :: (ToCBOR a, FromCBOR a) => Dual [a]
+dualList = Dual encodeFoldable (decodeList fromCBOR)
+
+dualSeq :: (ToCBOR a, FromCBOR a) => Dual (Seq a)
+dualSeq = Dual encodeFoldable (decodeSeq fromCBOR)
+
+dualSet :: (Ord a,ToCBOR a, FromCBOR a) => Dual (Set a)
+dualSet = Dual encodeFoldable (decodeSet fromCBOR)
 
-encodeList :: ToCBOR a => [a] -> Encoding
-encodeList = encodeFoldable
+dualMaybe :: (ToCBOR a, FromCBOR a) => Dual (Maybe a)
+dualMaybe = Dual toCBOR fromCBOR
```

Comment:
The encoding of `Maybe` provided by the `ToCBOR` instance encodes it as an empty or nonempty list. `[a] / []`.

The encoding we've been using is `a / null`.

Reply:
We made two different Duals for Maybe
dualMaybeAsList
dualMaybeAsNull